### ISSUE TITLE
cmd/cgo: C.* cannot be used for type constraints

### ISSUE THREAD

**Issue by hajimehoshi** [golang/go#52542]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/hajimehoshi/Library/Caches/go-build"
GOENV="/Users/hajimehoshi/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/hajimehoshi/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/hajimehoshi/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.18"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/hajimehoshi/test/generics/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/60/khbk2xqn1c5bml1byjn89dwc0000gn/T/go-build3847202091=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

Try compiling:

```go
package main

import "C"

func Foo[T ~C.int](x T) T {
        return x
}

func main() {
}
```

### What did you expect to see?

The compilation succeeds

### What did you see instead?

The compilation failed

```
$ go build main.go
# command-line-arguments
./main.go:5:13: undefined: C
```

**Comment by ianlancetaylor** [comment]
Rolling forward to 1.20.  Please comment if you disagree.  Thanks.

**Comment by mknyszek** [comment]
In triage, we think this should at least have a nice error message.

**Comment by mdempsky** [comment]
This will be fixed in 1.20. Note though that `~C.int` isn't a valid type parameter constraint, because `C.int` isn't its own underlying type.
